Provision alert notifications with YAML files

peay opened the pull request:
This addresses https://github.com/grafana/grafana/issues/10487, and allows to provision alert notifications just like dashboards and data sources using YAML configuration. Currently, one has to use the REST API manually which makes deployment workflows complicated.

Notes:
- each notification plugin has different settings, which would have to be documented for people to know how to configure them. I have not done that yet to first get some initial feedback.
- expanded `DeleteAlertNotification` into `DeleteAlertNotificationById`, `DeleteAlertNotificationByName` to support deleting alert notifications by name in the configuration.
- unlike the REST API (which can be tricked to accept any bogus alert notification `type`), the YAML configuration will validate the `type` field against existing notifiers. 

cc @bergquist

peay committed:
dev: provision alert notifications with YAML files

bergquist commented:
Great start! 

I wrote a small comment in the issue about how this should be solved in the long term. But I think this is a great start. I want to review it a little bit more before we merge it.

bergquist commented on pkg/services/provisioning/alert_notifications/types.go:
```diff
@@ -0,0 +1,116 @@
+	OrgId int64  `json:"org_id" yaml:"org_id"`
+}
+
+type notificationFromConfigV0 struct {
```
When we built provisioning for dashboards and datasources we didn't consider versions of the config files. So we had to retrofit in support for that. For alert notifiers, there is no point having multiple config structs. Yet :)

peay replied:
Yeah, I agree, wanted to stay consistent with the rest. Would you rather that I remove the config version completely right now? This can always be brought back if needed in the future.

peay commented:
@bergquist if you think we can move forward with this, I'll go ahead and document the parameter names for all notifiers.

peay commented:
@bergquist PR updated with:

- No API version
- Added the list of all the settings for all notifiers in the documentation. I've used the small script below to generate it, but I have not commit-ed it because it requires `x/net/html` which is currently not used by Grafana. Since notifiers are more or less frozen, I don't think this doc. will need to be updated very often, so it should be fine?

Side note: the notifiers are not very consistent about names of the settings. Some use camel case, some use underscore case, etc. This may be a bit confusing.

I will test this a bit more tomorrow.

```go
package main

import (
    "fmt"
    "strings"

    "github.com/grafana/grafana/pkg/components/simplejson"
    "github.com/grafana/grafana/pkg/services/alerting"
    "github.com/grafana/grafana/pkg/services/alerting/notifiers"
    "golang.org/x/net/html"
)

// This small script generates tables that summarize the settings for
// all notifiers. This can be used to populate `provisioning.md`.
//
// This inspects the HTML templates in each notifier to produce the full
// list.
func main() {
    notifiers.NewNotifierBase(1, false, "", "", simplejson.New())
    nts := alerting.GetNotifiers()

    for _, notifier := range nts {
        z := html.NewTokenizer(strings.NewReader(notifier.OptionsTemplate))

        fmt.Printf(`
#### Alert notification `+"`%s`"+`

| Name |
| ---- |
`, notifier.Type)

        done := false
        for !done {
            tt := z.Next()

            switch {
            case tt == html.ErrorToken:
                done = true
                break
            case tt == html.StartTagToken:
                t := z.Token()

                isField := t.Data == "input" || t.Data == "textarea"
                if isField {
                    isTextField := t.Data == "textarea"
                    settingName := ""

                    for _, attr := range t.Attr {
                        if attr.Key == "type" && attr.Val == "text" {
                            isTextField = true
                        } else if attr.Key == "ng-model" {
                            tokens := strings.Split(attr.Val, ".")
                            settingName = tokens[len(tokens)-1]
                        }
                    }

                    if isTextField {
                        fmt.Printf("| %s |\n", settingName)
                    }
                }
            }
        }
    }
}
```

peay commented:
I've been away for a while but I will try to get the last piece in this week (adding `uid` to alert notification channels)

Jokero commented:
@peay Could I help somehow?

vshiva commented:
ping !. Love to have this fix.  This is the only step that is currently manual is in our grafana/dashboard developer workflow automation.

wyhysj commented:
+1
would be nice if see in next release.

peay committed:
dev: add uid to alert notifications

peay commented:
I got a bit sidetracked, but I just rebased, fixed the conflicts, and had a first stab at adding the `uid`.

@bergquist This still needs more testing, but can you have a look to check if that's what you had in mind? Thanks!

Jokero commented:
@peay ping

peay commented:
@bergquist what do you think?

bergquist commented:
I'll take a look!

bergquist commented:
Great work!

Since the `id` field have been migrated to `uid` I think we should remove as much `id` usage as possible and only allow it in for legacy reasons in the dashboard.json. Right now both are kinda supported which I think will be harder to maintain. 

Seems you dropped the index for org_id and name but allow deletes/getby based on the name which will yield a weird behavior. Most queries for notifiers should use `uid`. The name should just be for display to make it easier read/rename.

ramato-procon commented:
@peay ping.. do you need help with @bergquist 's review?

wyhysj commented:
any updates on this

peay commented:
I am unfortunately unable to work on this at the moment - does someone want to pick it up?

apaz037 commented:
I would be interested in picking this up as I have a use case that requires this functionality.

yintothayang commented:
Thanks for picking this up @apaz037 , any updates?

banjoh commented:
Looking forward to getting this feature! Any updates?

pbakulev commented:
I've added implementation here https://github.com/grafana/grafana/pull/14229

bergquist commented:
closed by https://github.com/grafana/grafana/pull/14229
